core/store/src/trie/mod.rs

```diff
@@ -665,35 +713,100 @@ impl Trie {
+    fn nibbles_to_bytes(nibbles: &[u8]) -> (Vec<u8>, &[u8]) {
+        let (chunks, remainder) = stdx::as_chunks::<2, _>(nibbles);
+        let bytes = chunks.into_iter().map(|chunk| (chunk[0] * 16) + chunk[1]).collect::<Vec<u8>>();
+        (bytes, remainder)
+    }
+
     // Converts the list of Nibbles to a readable string.
     fn nibbles_to_string(prefix: &[u8]) -> String {
-        let (chunks, remainder) = stdx::as_chunks::<2, _>(prefix);
-        let mut result = chunks
-            .into_iter()
-            .map(|chunk| (chunk[0] * 16) + chunk[1])
-            .flat_map(|ch| std::ascii::escape_default(ch).map(char::from))
+        let (bytes, remainder) = Self::nibbles_to_bytes(prefix);
+        let mut result = bytes
+            .iter()
+            .flat_map(|ch| std::ascii::escape_default(*ch).map(char::from))
             .collect::<String>();
         if let Some(final_nibble) = remainder.first() {
             write!(&mut result, "\\x{:x}_", final_nibble).unwrap();
         }
         result
     }
 
+    /// Checks whether the provided `account_id_prefix` is lexicographically greater than `to` (if provided),
+    /// or whether it is lexicographically less than `from` (if provided, except being a prefix of `from`).
+    /// Although prefix of `from` is lexicographically less than `from`, pruning such subtree would cut off `from`.
+    fn is_out_of_account_id_bounds(
+        account_id_prefix: &[u8],
+        from: Option<&AccountId>,
+        to: Option<&AccountId>,
+    ) -> bool {
+        if let Some(from) = from {
+            if !from.as_bytes().starts_with(account_id_prefix)
+                && from.as_bytes() > account_id_prefix
+            {
+                return true;
+            }
+        }
+        if let Some(to) = to {
+            return account_id_prefix > to.as_bytes();
+        }
+        false
+    }
+
+    /// Returns true if the node with key `node_key` and its subtree should be skipped based on provided arguments.
+    /// If `record_type` is provided and the node is of different type, returns true.
+    /// If `AccountId`s in the subtree will not fall in the range [`from`, `to`], returns true.
+    /// Otherwise returns false.
+    fn should_prune_view_trie(
+        node_key: &Vec<u8>,
+        record_type: Option<u8>,
+        from: Option<&AccountId>,
+        to: Option<&AccountId>,
+    ) -> bool {
+        if node_key.is_empty() {
+            return false;
+        }
+
+        let column = node_key[0];
+        if let Some(record_type) = record_type {
+            if column != record_type {
+                return true;
+            }
+        }
+        if from.is_some() || to.is_some() {
+            if let Ok(account_id_prefix) = parse_account_id_prefix(column, &node_key) {
+                if Self::is_out_of_account_id_bounds(account_id_prefix, from, to) {
+                    return true;
+                }
+            }
+        }
```

Comment:
nice
the first if seems redundant given that is_out_of_account_id_bounds can handle options